src/backends/fvm_gpu.hpp

```diff
@@ -174,6 +185,129 @@ struct backend {
 
     static bool has_mechanism(const std::string& name) { return mech_map_.count(name)>0; }
 
+    /// threshold crossing logic
+    /// used as part of spike detection back end
+    class threshold_watcher {
+    public:
+        /// stores a single crossing event
+        struct threshold_crossing {
+            size_type index;    // index of variable
+            value_type time;    // time of crossing
+            __host__ __device__
+            friend bool operator==
+                (const threshold_crossing& lhs, const threshold_crossing& rhs)
+            {
+                return lhs.index==rhs.index && lhs.time==rhs.time;
+            }
+        };
+
+        using stack_type = gpu_stack<threshold_crossing>;
+
+        threshold_watcher() = default;
+
+        threshold_watcher(
+                const_view values,
+                const std::vector<size_type>& index,
+                const std::vector<value_type>& thresh,
+                value_type t=0):
+            values_(values),
+            index_(memory::make_const_view(index)),
+            thresholds_(memory::make_const_view(thresh)),
+            prev_values_(values),
+            is_spiking_(size()),
```

Comment:
The idea about using the `size()` function is that if you change the underlying data structure, or the way in which the "size" is computed in the future, you don't have to hunt through your code for all the locations where you used the old ad-hoc size calculation. Not that I have done that perfectly well everywhere...